cli/snapshot.go

```diff
@@ -0,0 +1,169 @@
+package cli
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// ioctl magics from <linux/fs.h>
+	IOCTL_FIFREEZE_MAGIC = 0xC0045877 // _IOWR('X', 119, int)
+	IOCTL_FITHAW_MAGIC   = 0xC0045878 // _IOWR('X', 120, int)
+)
+
+type SnapshotOptions struct {
+	ArtifactName      string
+	ArtifactGroup     string
+	CompatibleDevices []string
+	Depends           []string
+	Provides          []string
+}
+
+type mountInfo struct {
+	mountID    int
+	mountPoint string
+	mountSrc   string
+	fsType     string
+}
+
+func (runOpts *runOptionsType) DumpSnapshot(ctx *cli.Context) error {
+	rootDev, err := getFsDevFile("/")
+	if err != nil {
+		return err
+	}
+	dataDev, err := getFsDevFile(runOpts.dataStore)
+	if err != nil {
+		return err
+	}
+	if rootDev == dataDev {
+		return errors.Errorf(
+			"State data store (%s) is located on rootfs partition",
+			runOpts.dataStore)
+	}
+
+	errChan := make(chan error)
+	go func() {
+		thawChan := make(chan int)
+		sigChan := make(chan os.Signal)
+		signal.Notify(sigChan, unix.SIGINT, unix.SIGPIPE)
+		go func() {
+			select {
+			case <-sigChan:
+			case <-thawChan:
+			}
+			// FIXME: handle error
+			thawFs("/")
+			// Invoke os signal handler
+			signal.Stop(sigChan)
+			unix.Kill(os.Getpid(), unix.SIGINT)
+		}()
+		freezeFs("/")
+		unix.Sync()
+
+		f, err := os.Open(rootDev)
+		if err != nil {
+			thawChan <- 1
+			errChan <- err
+		}
+		defer f.Close()
+		io.Copy(os.Stdout, f)
+
+		thawChan <- 1
+		errChan <- nil
+		return
+	}()
+
+	err = <-errChan
+
+	return err
+}
+
+func getFsDevFile(fsRootPath string) (string, error) {
+	var statfs unix.Statfs_t
+	var stat unix.Stat_t
+
+	if err := unix.Statfs(fsRootPath, &statfs); err != nil {
+		return "", err
+	}
+
+	switch statfs.Type {
+	case unix.EXT4_SUPER_MAGIC:
+	case unix.REISERFS_SUPER_MAGIC:
+	case unix.XFS_SUPER_MAGIC:
+	case unix.JFFS2_SUPER_MAGIC:
+		// Filesystems that support FIFREEZE
+	default:
+		return "", errors.New("Filesystem not supported")
+	}
+	if err := unix.Stat(fsRootPath, &stat); err != nil {
+		return "", err
+	}
+
+	fsDevMajor := unix.Major(stat.Dev)
+	fsDevMinor := unix.Minor(stat.Dev)
+
+	linkPath, err := os.Readlink(
+		fmt.Sprintf("/dev/block/%d:%d", fsDevMajor, fsDevMinor))
+	if err != nil {
+		return "", errors.Wrap(err, "Error resolving device file path")
+	}
+	devPath := fmt.Sprintf("/dev/block/%s", linkPath)
```

Comment:
Suggest doing this instead of `Readlink` and `Sprintf`, since it won't work if `linkPath` is absolute:

```
devPath, err := filepath.EvalSymlinks(fmt.Sprintf("/dev/block/%d:%d", fsDevMajor, fsDevMinor))
```